### PR TITLE
[Inductor Cutlass backend] GEMM size threshold for Cutlass backend usage

### PR DESCRIPTION
**Pull request by kadeng**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121491
* #121490

 * Adds a configurable GEMM size threshold for the usage of Cutlass GEMM Kernels **_inductor.config.cutlass_backend_min_gemm_size**

 * During GEMM algorithm choice generation: **if no viable choices can be generated using the configured backends, the ATen backend will be used as a fallback backend**, even if it is not enabled in **_inductor.config.max_autotune_gemm_backends**

Test plan:
CI
Additional unit test in test_cutlass_backend.py

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/utils.py [hunk 1]
```diff
@@ -977,7 +977,15 @@ def use_triton_template(layout, *, enable_int32=False):
...
 
 
-def use_cutlass_template(layout):
+def use_cutlass_template(layout, m, n, k):
+    gemm_size = -1
+    try:
+        # this fails if it's dynamic
+        gemm_size = int(m * n * k)
+    except Exception:
+        pass
```
Comment (jansel) [inline torch/_inductor/utils.py:986]: Rather than erroring you can use `V.graph.sizevars.size_hint`.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/utils.py (hunk 1 above)
Comment (kadeng) [reply torch/_inductor/utils.py]: Changed that here. I think I should also use it in the PR above it in this stack ( about tolerating dynamic shapes ) to determine GEMM dimensions.

**Review by jansel — CHANGES_REQUESTED** [review]
Tests are failing

**Comment by kadeng** [comment]
> Tests are failing

Yes, I have been making the changes and introduced a circular import. That one's fixed and I'm waiting for CI.

File: torch/_inductor/utils.py [hunk 2]
```diff
@@ -977,7 +977,17 @@ def use_triton_template(layout, *, enable_int32=False):
...
 
-def use_cutlass_template(layout):
+def use_cutlass_template(layout, m, n, k):
+    from .virtualized import V
+
+    gemm_size = -1
+    try:
+        # this fails if it's dynamic
+        gemm_size = V.graph.sizevars.size_hint(m * n * k)
+    except Exception:
```
Comment (jansel) [inline torch/_inductor/utils.py:987]: Let's make this exception more narrow (if this is still needed).  This could hide bugs.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/utils.py (hunk 2 above)
Comment (kadeng) [reply torch/_inductor/utils.py]: makes sense, I eliminated the try catch entirely..

**Comment by kadeng** [comment]
Rebase introduced some new test failures, working on that..

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**